### PR TITLE
Add simple REPL functionality

### LINKED ISSUE #2: REPL example
It would be nice to have an interactive REPL in the `examples/` directory. 

- [x] Should not require any new dependencies, use standard IO.
- [x] Allow user to quit the REPL - check for `quit` (or any other keyword) before parsing it.
- [x] Allow user to toggle verbose mode on/off. 
- [x] Leverage the existing example for handling and printing errors.

### PR DESCRIPTION
**Pull request by NikolaJelic**
Adding simple REPL example that resolves #2 
Enable toggling `verbose` mode on and off using `-v` and listing commands using `-help`

### CODE REVIEW

File: examples/repl/repl.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include <kalcy/kalcy.hpp>
+#include <cassert>
+#include <format>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+namespace {
+struct Repl {
+	const std::string_view EXIT_TOKEN{"exit"};
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: `static constexpr`. (`const` members break assignment.)

Comment (karnkaul) [reply examples/repl/repl.cpp]: And not screaming case, instead `exit_token_v` or similar.

File: examples/repl/repl.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	bool verbose{}; // toggled on with '-v' and off with '-q'
+	bool is_running{true};
+
+	auto start() -> bool {
+		while (is_running) {
+			std::string text{};
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: Nitpick: for perfect consistency use `auto text = std::string{};`

File: examples/repl/repl.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			std::getline(std::cin, text);
+			// run kalcy on input expression
+			if (!run(text)) { return EXIT_FAILURE; }
+		}
+		// print epilogue
+		std::cout << std::format("\n^^ kalcy v{}\n", kalcy::version_v);
+		return EXIT_SUCCESS;
+	}
+
+	auto run(std::string_view const text) -> bool {
```
Comment (karnkaul) [inline examples/repl/repl.cpp:38]: Reduce nesting in this function. Leverage early returns to avoid `else if` / `else`.

File: examples/repl/repl.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		try {
+			// return false when the user enters the exit token
+			if (text == EXIT_TOKEN) {
+				is_running = false;
+			} else if (text == "-v") {
+				verbose = !verbose;
+			}  else {
+				// parse text into an expression
+				auto expr = kalcy::Parser{}.parse(text);
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: It makes sense to have a member `Parser` instead of creating a fresh one every time.

File: examples/repl/repl.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+				assert(expr != nullptr);
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: `expr` is expected to be null if the text passed is empty or just whitespace. This isn't possible with command line args, thus `quickstart` used an `assert`. 
Is that guaranteed here too?

File: examples/repl/repl.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+			return false;
+		}
+	}
+};
+} // namespace
+
+auto main(int argc, char** argv) -> int {
+
+	Repl repl{};
+	return repl.start();
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: Need program-level exception handling here.

```cpp
try {
	return repl.start();
} catch (std::exception const& e) {
  // ...
} catch (...) {
  // ...
}
```

File: examples/repl/repl.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	bool verbose{}; // toggled on with '-v' and off with '-q'
+	bool is_running{true};
+
+	auto start() -> bool {
+		while (is_running) {
+			std::string text{};
+			std::cout << std::format("[{}] > ", verbose ? "verbose" : "quiet");
+			std::getline(std::cin, text);
+			// run kalcy on input expression
+			if (!run(text)) { return EXIT_FAILURE; }
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: `return false`: `EXIT_*` is an `int`.

File: examples/repl/repl.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		}
+		// print epilogue
+		std::cout << std::format("\n^^ kalcy v{}\n", kalcy::version_v);
+		return EXIT_SUCCESS;
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: `return true`

File: examples/repl/repl.cpp (hunk 6 above)
Comment (karnkaul) [reply examples/repl/repl.cpp]: Also, `repl.start()` returns a `bool`, not `int`.

**Review by karnkaul — CHANGES_REQUESTED** [review]

File: examples/repl/repl.cpp (hunk 1 above)
Comment (NikolaJelic) [reply examples/repl/repl.cpp]: I have removed the token as a variable entirely after the last commit and am instead comparing `text` to the literal value.

File: examples/repl/repl.cpp (hunk 5 above)
Comment (NikolaJelic) [reply examples/repl/repl.cpp]: In the repl loop, `text` is checked before being passed to the `run` method and `run` won't be called on an empty string.

File: examples/repl/repl.cpp [hunk 9]
```diff
@@ -10,38 +10,58 @@ struct Repl {
 	const std::string_view EXIT_TOKEN{"exit"};
 	bool verbose{}; // toggled on with '-v' and off with '-q'
 	bool is_running{true};
+	kalcy::Parser parser{};
+
+	auto print_help() -> void {
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: Member function should be `const`

File: examples/repl/repl.cpp [hunk 10]
```diff
@@ -10,38 +10,58 @@ struct Repl {
...
 		while (is_running) {
-			std::string text{};
-			std::cout << std::format("[{}] > ", verbose ? "verbose" : "quiet");
+			auto text = std::string{};
+			std::cout << std::format("{} > ", verbose ? "[verbose]" : "");
 			std::getline(std::cin, text);
 			// run kalcy on input expression
-			if (!run(text)) { return EXIT_FAILURE; }
+			if (!text.empty()) {
+				if (!run(text)) { return EXIT_FAILURE; }
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: Don't return on failure, just restart the loop.

File: examples/repl/repl.cpp [hunk 11]
```diff
@@ -53,8 +73,10 @@ struct Repl {
 };
 } // namespace
 
-auto main(int argc, char** argv) -> int {
+auto main() -> int {
 
 	Repl repl{};
-	return repl.start();
+	try {
+		repl.start();
```
Comment (karnkaul) [inline examples/repl/repl.cpp:79]: If the return value of `start()` isn't even used, make it `void`?

File: kalcy/src/scanner.cpp [hunk 12]
```diff
@@ -24,8 +24,8 @@ auto Scanner::next() -> Token {
 	if (at_end()) { return {}; }
 
 	auto out = Token{};
-	if (match_number(out)) { return out; }
 	if (match_single(out)) { return out; }
+	if (match_number(out)) { return out; }
```
Comment (karnkaul) [inline kalcy/src/scanner.cpp]: Why are these changes here? Did you not rebase / merge with `origin/main`?

**Review by karnkaul — CHANGES_REQUESTED** [review]

File: kalcy/src/scanner.cpp (hunk 12 above)
Comment (NikolaJelic) [reply kalcy/src/scanner.cpp]: I have, and even the `-` bug is fixed.

File: examples/repl/repl.cpp [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	auto run(std::string_view const text) -> bool {
+		try {
+			if (text == "exit") {
+				is_running = false;
+				return true;
+			}
+
+			if (text == "-h" || text == "-help") {
```
Comment (karnkaul) [inline examples/repl/repl.cpp]: The long form of options is usually preceded by two dashes: `--help`, `--version`, etc.

**Review by karnkaul — APPROVED** [review]
